op-service/sources/receipts.go

```diff
@@ -4,18 +4,206 @@ import (
 	"context"
 	"fmt"
 	"io"
-	"sync"
+	"time"
 
 	"github.com/ethereum-optimism/optimism/op-service/sources/batching"
+	"github.com/ethereum-optimism/optimism/op-service/sources/caching"
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/common/hexutil"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/rpc"
 	"github.com/ethereum/go-ethereum/trie"
 
 	"github.com/ethereum-optimism/optimism/op-service/eth"
 )
 
+type ReceiptsProvider interface {
+	// FetchReceipts returns a block info and all of the receipts associated with transactions in the block.
+	// It verifies the receipt hash in the block header against the receipt hash of the fetched receipts
+	// to ensure that the execution engine did not fail to return any receipts.
+	FetchReceipts(ctx context.Context, block eth.BlockID, txHashes []common.Hash) (types.Receipts, error)
+}
+
+type CachingReceiptsProvider struct {
+	inner ReceiptsProvider
+	cache *caching.LRUCache[common.Hash, types.Receipts]
+}
```

Comment:
Nice. :)  I'm a big fan of having caching be a separate responsibility like this.